File: android/ValleTPV/app/src/main/java/com/valleapp/valletpv/dlg/DlgCobrar.java
package com.valleapp.valletpv.dlg;

import android.app.Dialog;
import android.content.Context;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.valleapp.valletpv.interfaces.IControladorCuenta;
import com.valleapp.valletpv.R;

import org.json.JSONArray;

import java.util.ArrayList;

/**
 * Created by valle on 19/10/14.
 */
public class DlgCobrar extends Dialog{

    JSONArray lineas;
    IControladorCuenta controlador;
    String strEntrega = "";
    Double totalCobro = 0.00;
    Double entrega = 0.00;
    TextView lblEntrega;
    TextView lblCambio;
    TextView lbltotal;

    public DlgCobrar(Context context, IControladorCuenta controlador) {
        super(context);
        this.controlador = controlador;
        setContentView(R.layout.cobros);
        lbltotal = findViewById(R.id.lblPrecio);
        lblEntrega =  findViewById(R.id.lblEntrega);
        lblCambio = findViewById(R.id.lblCambio);
        ImageButton tj = findViewById(R.id.btnTarjeta);
        ImageButton ef = findViewById(R.id.btnEfectivo);
        ImageButton s = findViewById(R.id.btn_salir_monedas);
        s.setOnClickListener(view -> clickSalir(view));
        ef.setOnClickListener(view -> clickEfectivo(view));

        tj.setOnClickListener(view -> clickTarjeta(view));
    }

    public void setDatos(JSONArray lineas, Double totalCobro){
        this.lineas = lineas;
        lbltotal.setText(String.format("%01.2f €", totalCobro));
        lblEntrega.setText(String.format("%01.2f €", totalCobro));
        this.totalCobro = totalCobro;
        this.entrega = totalCobro;
        this.strEntrega = "";
        LinearLayout pne = findViewById(R.id.pneBotonera);
        ArrayList<View> touchables = pne.getTouchables();

        for (View v : touchables){
            if (v instanceof Button){
                v.setOnClickListener(view -> clickEntrega(view));
            }
        }
    }


    public void clickEfectivo(View v){
        if(entrega>=totalCobro) {
            controlador.cobrar(lineas,totalCobro,entrega);
            this.cancel();
        }
    }

    public void clickTarjeta(View v){
        if (entrega == totalCobro) {
            controlador.cobrar(lineas, totalCobro, 0.00);
            this.cancel();
        }
    }

    public void clickSalir(View v){
        cancel();
    }

    public void clickEntrega(View v){
        String caracter = v.getTag().toString();
        if(caracter.equals("C")){
            entrega = totalCobro; strEntrega="";
            lblEntrega.setText(String.format("%01.2f €", totalCobro));
            lblCambio.setText("0.00 €");
        }else{
            try {
                strEntrega+=caracter;
                entrega = Double.parseDouble(strEntrega);
                lblEntrega.setText(String.format("%01.2f €", entrega));
                if(entrega>totalCobro)  lblCambio.setText(String.format("%01.2f €", entrega-totalCobro));
            }catch (Exception e){
                entrega = totalCobro; strEntrega= "";
                lblEntrega.setText(String.format("%01.2f €", totalCobro));
                lblCambio.setText("0.00 €");
            }
        }

    }

    @Override
    protected void onStop() {
        super.onStop();
        controlador.setEstadoAutoFinish(true, false);
    }
}

File: android/ValleTPV/ValleTPV/src/main/java/com/valleapp/valletpv/dlg/DlgCobrar.kt
package com.valleapp.valletpv.dlg

import android.app.Dialog
import android.content.Context
import android.view.View
import android.widget.Button
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.TextView
import com.valleapp.valletpv.R
import com.valleapp.valletpv.interfaces.IControladorCuenta
import org.json.JSONArray
import java.util.Locale

class DlgCobrar(
        context: Context,
        private val controlador: IControladorCuenta,
        private val usaCashlogy: Boolean,
        private val usaTPVPC: Boolean
) : Dialog(context) {

    private var lineas: JSONArray = JSONArray()
    private var totalCobro: Double = 0.00
    private var entrega: Double = 0.00
    private var strEntrega: String = ""
    private val lblEntrega: TextView
    private val lblCambio: TextView
    private val lblTotal: TextView

    init {
        setContentView(R.layout.cobros)

        // Evitar que el diálogo se cierre al tocar fuera
        setCanceledOnTouchOutside(false)

        // Evitar que se cierre con el botón "Atrás"
        setCancelable(false)


        lblTotal = findViewById(R.id.lblPrecio)
        lblEntrega = findViewById(R.id.lblEntrega)
        lblCambio = findViewById(R.id.lblCambio)
        val tj: ImageButton = findViewById(R.id.btnTarjeta)
        val ef: ImageButton = findViewById(R.id.btnEfectivo)
        val s: ImageButton = findViewById(R.id.btn_salir_monedas)
        val pne: LinearLayout = findViewById(R.id.pneBotonera)

        pne.visibility = if (usaCashlogy) View.GONE else View.VISIBLE

        s.setOnClickListener { clickSalir() }
        ef.setOnClickListener { clickEfectivo() }
        tj.setOnClickListener { clickTarjeta() }
    }

    fun setDatos(lineas: JSONArray, totalCobro: Double) {
        this.lineas = lineas
        this.totalCobro = totalCobro
        this.entrega = totalCobro
        this.strEntrega = ""

        lblTotal.text = String.format(Locale.getDefault(), "%01.2f €", totalCobro)
        lblEntrega.text = String.format(Locale.getDefault(), "%01.2f €", totalCobro)

        val pne: LinearLayout = findViewById(R.id.pneBotonera)

        for (i in 0 until pne.childCount) {
            val layout = pne.getChildAt(i)
            if (layout is LinearLayout) {
                for (j in 0 until layout.childCount) {
                    val btn = layout.getChildAt(j)
                    if (btn is Button) {
                        btn.setOnClickListener { clickEntrega(btn) }
                    }
                }
            }
        }
    }

    private fun clickEfectivo() {
        if (usaCashlogy) {
            controlador.cobrarConCashlogy(lineas, totalCobro)
            clickSalir()
            return
        }
        if (entrega >= totalCobro) {
            clickSalir()
            controlador.cobrar(lineas, totalCobro, entrega, "")
        }
    }

    private fun clickTarjeta() {
        if (usaTPVPC) {
            controlador.cobrarConTpvPC(lineas, totalCobro)
            clickSalir()
            return
        }
        if (entrega == totalCobro) {
            clickSalir()
            controlador.cobrar(lineas, totalCobro, 0.00, "")
        }
    }

    private fun clickSalir() {
        cancel()
    }

    private fun clickEntrega(v: View) {
        val caracter = v.tag.toString()
        if (caracter == "C") {
            entrega = totalCobro
            strEntrega = ""
            lblEntrega.text = String.format(Locale.getDefault(), "%01.2f €", totalCobro)
            lblCambio.text = "0.00 €"
        } else {
            try {
                strEntrega += caracter
                entrega = strEntrega.toDouble()
                lblEntrega.text = String.format(Locale.getDefault(), "%01.2f €", entrega)
                if (entrega > totalCobro) {
                    lblCambio.text = String.format(Locale.getDefault(), "%01.2f €", entrega - totalCobro)
                }
            } catch (e: Exception) {
                entrega = totalCobro
                strEntrega = ""
                lblEntrega.text = String.format(Locale.getDefault(), "%01.2f €", totalCobro)
                lblCambio.text = "0.00 €"
            }
        }
    }

    override fun onStop() {
        controlador.setEstadoAutoFinish(true, false)
        super.onStop()
    }
}